OpenShift CLI cmd for volumes

pravisankar opened the pull request:
closes #2643

ncdc commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,588 @@
+
+	volumeExample = `  // Add new volume of type 'emptyDir' for deployment 'registry' and mount under /opt inside the containers
+  $ %[1]s volume dc/registry --add --mount-path=/opt
+  Here name of the volume is auto generated and default type 'emptyDir' will be used.
```
Move this up above the command, prefix with `//`: `// The volume name is auto generated`. I don't think you need to mention `emptyDir` again as it's in the 1st comment line

ncdc commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,588 @@
+  $ %[1]s volume pod/p1 --add -m /repo --source=<json-string>
+
+  // Change rc 'r1' to use persistent volume instead of emptyDir (vol name: v1)
+  $ %[1]s volume rc/r1 --add --name=v1 -t persistentVolumeClaim --claim-name=pvc1 --confirm
```
So "update" is an "add" with an existing name + --confirm?

pravisankar replied:
yes

ncdc replied:
I don't like that. Can we either rename add to set, or distinguish between add and update?

ncdc replied:
@smarterclayton

smarterclayton replied:
--confirm is about overwriting an existing volume.  If you say "I want to add volume Foo to Bar, and Foo already exists", "--confirm" seems like the natural way to go past that.  Update implies changing the thing that exists, but what we're doing here is a rip and replace.

ncdc replied:
But you're not adding. You're removing and then re-adding. The English speaker in me thinks add/update or set are more appropriate, but I won't hold this up if everyone else is fine with add+confirm=rip+replace.

deads2k replied:
If it's a vote, I'm pro-"set".

smarterclayton replied:
This is mostly modeled after env, which is an implicit set with no --confirm needed.  In this case, if --add is going to fail if something already exists, it's not --update.  --update would be "i'm going to change something that exists".  --set implies you don't need a --confirm to me.

ncdc replied:
So this doesn't really get us to a good UX yet.

I am really hung up on add meaning n++. And that's why add+confirm for an existing volume where n doesn't increase seems really weird to me.

I agree update(nonexisting) also seems weird.

That's why I like set. But I can see where set shouldn't need confirm. How about set + overwrite to confirm/force an update?

ncdc replied:
I'd probably be ok with add+overwrite if set+overwrite isn't desirable.

smarterclayton replied:
Yeah, it should be overwrite - that's what we use on env/labels/create-keys commands.  --confirm is more generic.

----- Original Message -----
> > +any object that has a pod template (replication controllers or deployment
> > configurations).
> > +You can specify a single object or multiple, and alter volumes on all
> > containers or
> > +just those that match a wildcard.`
> > +
> > +	volumeExample = `  // Add new volume of type 'emptyDir' for deployment
> > 'registry' and mount under /opt inside the containers
> > +  $ %[1]s volume dc/registry --add --mount-path=/opt
> > +  Here name of the volume is auto generated and default type 'emptyDir'
> > will be used.
> > +
> > +  // Add new volume 'v1' with secret 'magic' for pod 'p1'
> > +  $ %[1]s volume pod/p1 --add --name=v1 -m /etc --type=secret
> > --secret-name=magic
> > +
> > +  // Add new volume to pod 'p1' based on gitRepo (or other volume sources
> > not supported by --type)
> > +  $ %[1]s volume pod/p1 --add -m /repo --source=<json-string>
> > +
> > +  // Change rc 'r1' to use persistent volume instead of emptyDir (vol
> > name: v1)
> > +  $ %[1]s volume rc/r1 --add --name=v1 -t persistentVolumeClaim
> > --claim-name=pvc1 --confirm
> 
> I'd probably be ok with add+overwrite if set+overwrite isn't desirable.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/3023/files#r32159778

smarterclayton replied:
Just change the example text - "Create a new persistent volume and overwrite the existing volume"

ncdc commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,588 @@
+  $ %[1]s volume rc/r1 --add --name=v1 -t persistentVolumeClaim --claim-name=pvc1 --confirm
+
+  // Change pod 'p1' mount point to /data for volume v1
+  $ %[1]s volume pod p1 --add --name=v1 -m /reg --confirm
```
data or reg?

pravisankar replied:
yup, need to be consistent with comment

ncdc commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,588 @@
+  $ %[1]s volume pod/p1 v --remove --name=v1 --containers=c1
+
+  // List volumes defined on replication controller 'r1'
+  $ %[1]s volume rc r1 --list
```
does this need to be rc/r1?

pravisankar replied:
'rc r1' or 'rc/r1' both notations work

ncdc commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,588 @@
+  // Remove all volumes for pod 'p1'
+  $ %[1]s volume pod/p1 --remove --confirm
+
+  // Remove volume 'v1' from deployment 'registry'  
```
deployment config

ncdc commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,588 @@
+You can specify a single object or multiple, and alter volumes on all containers or
+just those that match a wildcard.`
+
+	volumeExample = `  // Add new volume of type 'emptyDir' for deployment 'registry' and mount under /opt inside the containers
```
deployment config

pravisankar commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,588 @@
+			errList = append(errList, errors.New("--path|--secret-name|--claim-name are only valid for --type option"))
+		}
+
+		if len(a.Source) != 1 {
```
len(a.Source) > 1

ncdc commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,588 @@
+		numOps++
+	}
+	if numOps == 0 {
+		errList = append(errList, errors.New("must provide a volume operation. Support operations: --add, --remove and --list"))
```
you must specify an operation; valid values are --add, --remove, and --list

ncdc commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,588 @@
+	if numOps == 0 {
+		errList = append(errList, errors.New("must provide a volume operation. Support operations: --add, --remove and --list"))
+	} else if numOps > 1 {
+		errList = append(errList, errors.New("must provide only one volume operation at a time"))
```
you may only specify one operation at a time

ncdc commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,588 @@
+	return kerrors.NewAggregate(errList)
+}
+
+func (a *AddVolumeOptions) Validate(isAddOp bool) error {
```
This method is pretty big, has a lot of nested conditionals, and very little whitespace. Could you at least put some whitespace in between logical units, if nothing else?

ncdc commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,588 @@
+	}
+
+	// only apply resource version locking on a single resource
+	if !one && len(v.ResourceVersion) > 0 {
```
You don't appear to do anything with v.ResourceVersion other than make sure it's only used with 1 resource. Also, is the intended use of this that you'd `oc get <resource>`, make note of its resource version, then run `oc volume --resource-version=...`?

smarterclayton replied:
Let's just drop that feature, it's not very useful right now and I'd rather not get sidetracked

> On Jun 9, 2015, at 8:17 PM, Andy Goldstein <notifications@github.com> wrote:
> 
> In pkg/cmd/cli/cmd/volume.go:
> 
> > +func (v *VolumeOptions) RunVolume(args []string) error {
> > +	b := resource.NewBuilder(v.Mapper, v.Typer, resource.ClientMapperFunc(v.RESTClientFactory)).
> > +		ContinueOnError().
> > +		NamespaceParam(v.DefaultNamespace).DefaultNamespace().
> > +		SelectorParam(v.Selector).
> > +		ResourceTypeOrNameArgs(v.All, args...).
> > +		Flatten()
> > +
> > +	one := false
> > +	infos, err := b.Do().IntoSingular(&one).Infos()
> > +	if err != nil {
> > +		return err
> > +	}
> > +
> > +	// only apply resource version locking on a single resource
> > +	if !one && len(v.ResourceVersion) > 0 {
> You don't appear to do anything with v.ResourceVersion other than make sure it's only used with 1 resource. Also, is the intended use of this that you'd oc get <resource>, make note of its resource version, then run oc volume --resource-version=...?
> 
> —
> Reply to this email directly or view it on GitHub.
>

pravisankar commented:
@ncdc @fabianofranz @smarterclayton  updated, PTAL

fabianofranz commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,596 @@
+/*
```
Remove this copyright section

fabianofranz commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,596 @@
+	if v.List {
+		numOps++
+	}
+	if numOps == 0 {
```
switch numOps

fabianofranz commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,596 @@
+	for _, info := range infos {
+		ok, err := v.UpdatePodSpecForObject(info.Object, func(spec *kapi.PodSpec) error {
+			var e error
+			if v.Add {
```
switch

fabianofranz commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,596 @@
+	for _, info := range infos {
+		data, err := info.Mapping.Codec.Encode(info.Object)
+		if err != nil {
+			fmt.Fprintf(v.Writer, "Error: %v\n", err)
```
Instead of printing the error here and using the `failed` control, you could use an `err` array and `errors.NewAggregate(errs)`.

liggitt commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,596 @@
+	addOpts := &AddVolumeOptions{}
+	opts := &VolumeOptions{AddOpts: addOpts}
+	cmd := &cobra.Command{
+		Use:     "volume RESOURCE/NAME --add|--remove|--list",
```
shouldn't these be subcommands?

liggitt replied:
if you find yourself having lots of boolean options that are mututally exclusive, or flags which only apply in a particular mode, you probably want subcommands

ncdc replied:
We discussed this on a scrum the other day. My preference was subs, Clayton said it would depend on how many common flags there are. I think looking at the examples, subs makes more sense.

smarterclayton replied:
I don't consider three too many.  One of those is the default, so it's really 2.

----- Original Message -----
> > +type AddVolumeOptions struct {
> > +	Type       string
> > +	MountPath  string
> > +	Path       string
> > +	SecretName string
> > +	ClaimName  string
> > +	Source     string
> > +}
> > +
> > +var typeChanged bool
> > +
> > +func NewCmdVolume(fullName string, f *clientcmd.Factory, out io.Writer)
> > *cobra.Command {
> > +	addOpts := &AddVolumeOptions{}
> > +	opts := &VolumeOptions{AddOpts: addOpts}
> > +	cmd := &cobra.Command{
> > +		Use:     "volume RESOURCE/NAME --add|--remove|--list",
> 
> We discussed this on a scrum the other day. My preference was subs, Clayton
> said it would depend on how many common flags there are. I think looking at
> the examples, subs makes more sense.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/3023/files#r32127842

ncdc replied:
I think @liggitt is referring to the options that vary between add, remove, list.

pravisankar replied:
Except Type, Path, SecretName and ClaimName, rest of the falgs are all common for the operations (~60 to 70%).
We had similar situation with 'oadm manage-node' and the recommendation was to have singe cmd: avoids code duplication and user don't need to browse couple of man/help pages to execute a cmd.

fabianofranz replied:
Is one of `--add|--remove|--list` always required? If not, the usage here should be `[--add|--remove|--list]`.

pravisankar replied:
Yes, user need to specify one of the operation

pravisankar commented:
Updated, I've tested the common cases on my local machine. Please review/merge

ncdc commented:
I still think we either need `--add` and `--update` as distinct operations, or we just use `--set`. add+confirm=update seems wrong.

pravisankar commented:
Added '--overwrite' option for replacing volume/volume-mounts. Please review/merge
Later I'll update hack/test-cmd.sh with the volume test cases.

ncdc commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,585 @@
+		}
+	}
+	if len(args) < 1 {
+		errList = append(errList, errors.New("one ore more resources must be specified as <resource> <name> or <resource>/<name>"))
```
s/ore/or

ncdc commented on pkg/cmd/cli/cli.go:
```diff
@@ -97,6 +97,8 @@ func NewCommandCLI(name, fullName string) *cobra.Command {
 				cmd.NewCmdDescribe(fullName, f, out),
 				cmd.NewCmdEdit(fullName, f, out),
 				cmd.NewCmdEnv(fullName, f, os.Stdin, out),
+				// TODO: Enable this cmd and remove from experimental(post 3.0 release)
```
Enable this

ncdc commented on pkg/cmd/openshift/openshift.go:
```diff
@@ -148,6 +148,7 @@ func newExperimentalCommand(name, fullName string) *cobra.Command {
 
 	f := clientcmd.New(experimental.PersistentFlags())
 
+	experimental.AddCommand(cmd.NewCmdVolume(fullName, f, out))
```
Remove from experimental

ncdc commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,585 @@
+	Source     string
+}
+
+var typeChanged bool
```
Took me a little while to figure out what this is doing. Looks a little weird as a package global.

smarterclayton replied:
No package globals

----- Original Message -----
> > +
> > +	// Add op params
> > +	AddOpts *AddVolumeOptions
> > +}
> > +
> > +type AddVolumeOptions struct {
> > +	Type       string
> > +	MountPath  string
> > +	Overwrite  bool
> > +	Path       string
> > +	SecretName string
> > +	ClaimName  string
> > +	Source     string
> > +}
> > +
> > +var typeChanged bool
> 
> Took me a little while to figure out what this is doing. Looks a little weird
> as a package global.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/3023/files#r32255991

ncdc commented:
It might be nice to write to cmd.Out() the generated name of the volume, if the user doesn't specify --name

ncdc commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,585 @@
+			if !opts.Overwrite {
+				return fmt.Errorf("Volume '%s' already exists. Use --overwrite to replace", v.Name)
+			}
+			if !typeChanged && len(opts.Source) == 0 {
```
I've been staring at this for a while trying to figure it out. Is this the correct interpretation: when doing --overwrite, if you didn't specify --type, and if you didn't specify --source, it means you're doing the default --type=emptyDir, in which case you copy the existing VolumeSource to newVolume, remove the existing volume from spec.Volumes, add it back below, and maybe change the mount point?

pravisankar replied:
yes

ncdc replied:
Would it be possible to call the function to update the VolumeMounts (https://github.com/openshift/origin/pull/3023/files#r32259036) and just return early? No need to remove from spec.Volumes only to re-add.

ncdc replied:
And setSource can then be removed as it wouldn't be needed any more

pravisankar replied:
If we don't remove and re-add the volume in case of overwrite, we need to unset the existing volume and I don't want to take that route (reason: we may not be able to check all available volume sources assuming there will be addition of new volume sources in the future).

ncdc commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,585 @@
+func (v *VolumeOptions) setVolumeSource(kv *kapi.Volume) error {
+	opts := v.AddOpts
+	if len(opts.Type) > 0 {
+		switch strings.ToLower(opts.Type) {
```
make this its own function for updating the Volume based on type

ncdc commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,585 @@
+	spec.Volumes = append(spec.Volumes, *newVolume)
+
+	if len(opts.MountPath) > 0 {
+		containers, _ := v.selectContainers(spec.Containers, v.Containers)
```
I'd make this a separate function

fabianofranz commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,602 @@
+  $ %[1]s volume pod p1 --add --name=v1 -m /data --overwrite
+
+  // Remove all volumes for pod 'p1'
+  $ %[1]s volume pod/p1 --remove --confirm
```
I know there was some discussions around `--confirm` already, but: what happens if you don't specify `--confirm` here?

pravisankar replied:
Removal of more than one volume needs '---confirm' flag otherwise we will throw an error and will recommend user to retry the op with --confrim flag.

smarterclayton commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,602 @@
+// selectString returns true if the provided string matches spec, where spec is a string with
+// a non-greedy '*' wildcard operator.
+// TODO: turn into a regex and handle greedy matches and backtracking.
+func (v *VolumeOptions) selectString(s, spec string) bool {
```
Why doesn't this reuse the other select string function

smarterclayton commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,602 @@
+	}
+
+	if !errored {
+		fmt.Fprintf(v.Writer, "Error: %v\n", err)
```
error:

smarterclayton commented on pkg/cmd/cli/cmd/volume.go:
```diff
@@ -0,0 +1,602 @@
+	return match
+}
+
+func (v *VolumeOptions) handleUpdateError(err error) {
```
This method shouldn't be part of volumeOptions because it doesn't need any local fields.

pravisankar replied:
It needs v.Writer, may be i can pass it as an argument

smarterclayton replied:
Generally yes.

----- Original Message -----
> > +			fallthrough
> > +		// first part does not match start of string
> > +		case i == 0 && pos != 0:
> > +			fallthrough
> > +		// last part does not exactly match remaining part of string
> > +		case i == (len(parts)-1) && len(s) != (len(part)+next):
> > +			match = false
> > +			break
> > +		default:
> > +			pos = next
> > +		}
> > +	}
> > +	return match
> > +}
> > +
> > +func (v *VolumeOptions) handleUpdateError(err error) {
> 
> It needs v.Writer, may be i can pass it as an argument
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/3023/files#r32336497

pravisankar commented:
Updated, Moved common code in env.go and  volume.go to pkg/cmd/util/resource.go
Tested the cmds one more time on local machine. 
In today's demo, I type -name instead of --name and that caused weird errors like replication controller not found. Understood the problem, -name was treated as -n ame and it couldn't find the replication controller for namespace 'ame'!

fabianofranz commented:
@pravisankar latest version LGTM.

smarterclayton commented on pkg/cmd/util/resource.go:
```diff
@@ -0,0 +1,80 @@
+package util
```
We generally shouldn't be adding these things to cmd/util - it has nothing to do with those.  Instead, extract it within pkg/cmd/cli/cmd/* as non public methods.

pravisankar commented:
Common methods for env and volume removed from cmd/util and moved under pkg/cmd/cli/cmd/helper.go as non public methods.
@smarterclayton please merge if you are okay with the changes

smarterclayton commented:
[merge]
